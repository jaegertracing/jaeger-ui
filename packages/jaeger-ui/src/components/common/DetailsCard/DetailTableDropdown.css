/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DetailTableDropdown--Footer {
  background: white;
  border: 1px solid #ccc;
  display: flex;
  justify-content: space-between;
  padding: 0.3em;
}

.DetailTableDropdown--Footer--CancelConfirm {
  display: flex;
}

.DetailTableDropdown--Btn {
  align-items: center;
  border-radius: 0;
  color: #ddd;
  display: flex;
  padding: 0.3em 0.7em 0.3em 0.6em;
}

.DetailTableDropdown--Btn ~ .DetailTableDropdown--Btn {
  margin-left: 0.3em;
}

.DetailTableDropdown--Btn > *:not(:first-child) {
  margin-left: 0.3em;
}

.DetailTableDropdown--Btn.Apply {
  background: #4c21ce;
}

.DetailTableDropdown--Btn.Cancel {
  background: #007272;
}

.DetailTableDropdown--Btn.Clear {
  background: #ff2626;
}

.DetailTableDropdown--Tooltip {
  max-width: unset;
  white-space: nowrap;
}

.DetailTableDropdown--Tooltip--Body {
  display: flex;
  flex-direction: column;
}
