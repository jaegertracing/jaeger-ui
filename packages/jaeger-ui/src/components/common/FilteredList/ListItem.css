/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.FilteredList--ListItem {
  align-items: center;
  cursor: pointer;
  display: flex;
  padding: 0.5em 5em 0.5em 0.5em;
  white-space: nowrap;
  min-width: 100%;
}

.FilteredList--ListItem.is-striped {
  background: linear-gradient(to right, rgba(0, 0, 0, 0.04), rgba(0, 0, 0, 0));
}

.FilteredList--ListItem.is-striped:hover,
.FilteredList--ListItem:hover {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.1),
    rgba(135, 232, 222, 0.2) 70%,
    rgba(135, 232, 222, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-selected {
  background: linear-gradient(to right, #87e8de, #87e8deaa 70%, #87e8de00);
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-striped.is-focused,
.FilteredList--ListItem.is-focused {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.2),
    rgba(17, 153, 153, 0.2) 70%,
    rgba(17, 153, 153, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem mark {
  background: none;
  color: #eb2f96;
  font-weight: 500;
  padding: 0;
}
