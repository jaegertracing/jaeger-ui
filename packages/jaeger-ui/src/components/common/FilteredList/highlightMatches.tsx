// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import memoize from 'lru-memoize';

type TIsHighlightGroupFn = (index: number) => boolean;

type THighlightMatchFn = (match: RegExpMatchArray) => React.ReactNode;

function toHighlights(isHighlightGroup: TIsHighlightGroupFn, match: RegExpMatchArray) {
  const texts = match
    .map((tx: string, i: number) => {
      if (i === 0 || !tx) {
        // the first group is the full match, not a capturing group
        return null;
      }
      if (isHighlightGroup(i)) {
        return <mark key={`${tx + i}`}>{tx}</mark>;
      }
      return tx;
    })
    .filter(Boolean);
  return <span>{texts}</span>;
}

// In the matches from the regular expressions created in `MatchHighlighter`,
// text that should be highlighted is interleaved with text that shouldn't be
// highlighted. E.g. alternating capturing groups should be highlighted,
// regardless of how many capturing groups there are in the match. Therefore, we
// need to establish if the first capturing group (i.e. match[1]) should be
// highlighted or if it should start at match[2]. Then, we alternate, from there.
// This results in either highlighting even groups (starting at match[2]) or odd
// groups (starting at match[1]). However, we don't actually need to highlight
// odd groups, but just the first group (match[1]), so we can simplify that test.
const highlightEvenGroups = toHighlights.bind(null, (index: number) => index % 2 === 0);
const highlightFirstGroup = toHighlights.bind(null, (index: number) => index === 1);

const ANY_LAZY = '(.*?)';
const ANY_GREEDY = '(.*)';

function wordBreak(lowerLetter: string) {
  return `(\\b${lowerLetter}|${lowerLetter.toUpperCase()})`;
}

function letterClass(lowerLetter: string) {
  return `[${lowerLetter}${lowerLetter.toUpperCase()}]`;
}

/**
 * Create regular expressions used for highlighting matches between the query
 * and the text. This is derived from the match-sorter ranking system description:
 * https://github.com/kentcdodds/match-sorter#this-solution
 *
 * Several regular expressions are created from the filterText. These roughly
 * align with the ranking system described in the link. They are used to divide
 * input text into emphasized and non-emphasized sections.
 */
class MatchHighlighter {
  public static make(query: string) {
    return new MatchHighlighter(query);
  }

  private readonly matchers: [RegExp, THighlightMatchFn][];

  constructor(readonly query: string) {
    // istanbul ignore next
    if (!query) {
      this.matchers = [];
      return;
    }
    const tx = query.toLowerCase();
    const initialCap = `${tx[0].toUpperCase()}${tx.slice(1)}`;
    const letters = tx.split('');
    const startsWith = `^(${tx})`;
    const wordStartsWith = `(\\b${letters.map(letterClass).join('')}|${initialCap})`;
    const acronymLetters = letters.map(letter => wordBreak(letter)).join(ANY_LAZY);
    const constains = `(${tx})`;
    const anyLetters = letters.map(letter => `(${letter})`).join(ANY_LAZY);

    // given the query "eg", the following regular expressions will be created:
    this.matchers = [
      // case-insensitive match at start of text, would match "ego" but not "lego"
      // /^(eg)(.*)/i
      [new RegExp(`${startsWith}${ANY_GREEDY}`, 'i'), highlightFirstGroup],

      // case-insensitive match at start of word, with boundaries in camelCase counting as words
      // would match "theEgg" and "the-egg"
      // /(.*?)(\b[eE][gG]|Eg)(.*)/
      [new RegExp(`${ANY_LAZY}${wordStartsWith}${ANY_GREEDY}`), highlightEvenGroups],

      // match an acronym, e.g. "easy-going" and "easyGoing" would both match "eg"
      // /(.*?)(\be|E)(.*?)(\bg|G)(.*)/
      [new RegExp(`${ANY_LAZY}${acronymLetters}${ANY_GREEDY}`), highlightEvenGroups],

      // case insensitive contains
      // /(.*?)(eg)(.*)/i
      [new RegExp(`${ANY_LAZY}${constains}${ANY_GREEDY}`, 'i'), highlightEvenGroups],

      // contains each letter, in sequence
      // /(.*?)(e)(.*?)(g)(.*)/i
      [new RegExp(`${ANY_LAZY}${anyLetters}${ANY_GREEDY}`, 'i'), highlightEvenGroups],
    ];
  }

  highlightMatches(text: string) {
    if (!text) {
      return text;
    }
    for (let i = 0; i < this.matchers.length; i++) {
      const [matcher, highlightMatch] = this.matchers[i];
      const match = text.match(matcher);
      if (match) {
        return highlightMatch(match);
      }
    }
    return text;
  }
}

const getHighlighter = memoize(30)(MatchHighlighter.make);

function highlightMatchesImpl(query: string, text: string) {
  if (!query) {
    return text;
  }
  return getHighlighter(query).highlightMatches(text);
}

export default memoize(200)(highlightMatchesImpl);
