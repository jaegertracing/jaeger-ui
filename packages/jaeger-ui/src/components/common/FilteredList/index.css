/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.FilteredList--list {
  background: #fafafa;
}

.FilteredList--filterWrapper {
  align-items: center;
  display: flex;
  box-shadow: 0 0 3px 1px rgba(0, 0, 0, 0.3);
  position: relative;
  z-index: 10;
}

.FilteredList--filterIcon {
  font-size: 1.3em;
  margin: 0 0.25em 0 0.65em;
  position: absolute;
}

.FilteredList--filterInput {
  border: none;
  flex: 1;
  height: auto;
  padding: 0.5em 0.3em 0.5em 3em;
}

.FilteredList--filterInput:focus {
  box-shadow: inset 0px 0px 3px 1px #6aa;
  outline: 1px solid #199;
  outline-offset: -1px;
}
