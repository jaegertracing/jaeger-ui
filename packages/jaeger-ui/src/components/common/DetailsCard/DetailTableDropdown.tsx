// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button, Tooltip } from 'antd';
import FaCheck from 'react-icons/lib/fa/check.js';
import FaTrash from 'react-icons/lib/fa/trash.js';
import TiCancel from 'react-icons/lib/ti/cancel.js';

import FilteredList from '../FilteredList';

import { TFilterDropdownProps } from './types';

import './DetailTableDropdown.css';

type TProps = TFilterDropdownProps & {
  options: Set<string>;
};

export default class DetailTableDropdown extends React.PureComponent<TProps> {
  cancelled = false;
  selected: React.Key[] = [];

  componentDidUpdate(prevProps: TProps) {
    const { confirm, selectedKeys } = this.props;

    // If the entries in selectedKeys is unchanged, the dropdown has opened or closed.
    // Record the selectedKeys at this time for future cancellations.
    if (selectedKeys.length === prevProps.selectedKeys.length) {
      const prevKeys = new Set(prevProps.selectedKeys);
      if (selectedKeys.every(key => prevKeys.has(key))) {
        this.selected = selectedKeys;
      }
    }

    // Unfortunately antd requires setSelectedKeys and confirm to be called in different cycles.
    if (this.cancelled) {
      this.cancelled = false;
      confirm();
    }
  }

  cancel = () => {
    // Unfortunately antd requires setSelectedKeys and confirm to be called in different cycles.
    this.cancelled = true;
    this.props.setSelectedKeys(this.selected);
  };

  render() {
    const { clearFilters = () => {}, confirm, options, selectedKeys, setSelectedKeys } = this.props;

    const value = new Set<string>();
    selectedKeys.forEach(selected => {
      if (typeof selected === 'string') value.add(selected);
    });

    return (
      <div>
        <FilteredList
          addValues={(values: string[]) => {
            setSelectedKeys([...selectedKeys, ...values]);
          }}
          multi
          options={Array.from(options)}
          removeValues={(values: string[]) => {
            const remove = new Set<React.Key>(values);
            setSelectedKeys(selectedKeys.filter(key => !remove.has(key)));
          }}
          setValue={(v: string) => {
            setSelectedKeys([v]);
          }}
          value={value}
        />
        <div className="DetailTableDropdown--Footer">
          <Tooltip overlayClassName="DetailTableDropdown--Tooltip" title="Remove filter from this column">
            <Button className="DetailTableDropdown--Btn Clear" onClick={clearFilters}>
              <FaTrash size={18} />
              Clear Filter
            </Button>
          </Tooltip>
          <div className="DetailTableDropdown--Footer--CancelConfirm">
            <Tooltip
              overlayClassName="DetailTableDropdown--Tooltip"
              title="Cancel changes to this column's filter"
            >
              <Button className="DetailTableDropdown--Btn Cancel" onClick={this.cancel}>
                <TiCancel size={20} />
                Cancel
              </Button>
            </Tooltip>
            <Tooltip
              overlayClassName="DetailTableDropdown--Tooltip"
              title={
                <div className="DetailTableDropdown--Tooltip--Body">
                  <span>Apply changes to this column{"'"}s filter</span>
                  <span>Same effect as clicking outside the dropdown</span>
                </div>
              }
            >
              <Button className="DetailTableDropdown--Btn Apply" onClick={confirm}>
                <FaCheck size={18} />
                Apply
              </Button>
            </Tooltip>
          </div>
        </div>
      </div>
    );
  }
}
