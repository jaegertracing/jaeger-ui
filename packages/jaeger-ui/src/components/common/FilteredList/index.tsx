// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox, Tooltip } from 'antd';
import _debounce from 'lodash/debounce';
import matchSorter from 'match-sorter';
import IoIosSearch from 'react-icons/lib/io/ios-search';
import { FixedSizeList as VList, ListOnItemsRenderedProps, ListOnScrollProps } from 'react-window';
import { Key as EKey } from 'ts-key-enum';

import ListItem from './ListItem';

import './index.css';

const ITEM_HEIGHT = 35;
const MAX_HEIGHT = 375;

type TProps = {
  addValues?: (values: string[]) => void;
  cancel?: () => void;
  multi?: boolean;
  options: string[];
  removeValues?: (values: string[]) => void;
  setValue: (value: string) => void;
  value: Set<string> | string | null;
};

type TState = {
  filterText: string;
  visibleStartIndex: number;
  visibleStopIndex: number;
  focusedIndex: number | null;
};

export default class FilteredList extends React.PureComponent<TProps, TState> {
  inputRef: React.RefObject<HTMLInputElement> = React.createRef();
  vlistRef: React.RefObject<VList> = React.createRef();
  wrapperRef: React.RefObject<HTMLDivElement> = React.createRef();
  state: TState = {
    filterText: '',
    focusedIndex: null,
    visibleStartIndex: 0,
    visibleStopIndex: 0,
  };

  componentDidUpdate() {
    this.focusInput();
  }

  focusInput = () => {
    const { current } = this.inputRef;
    if (current) {
      current.focus();
    }
  };

  isMouseWithin() {
    /* istanbul ignore next */
    const { current } = this.wrapperRef;
    /* istanbul ignore next */
    return current != null && current.matches(':hover');
  }

  private getFilteredCheckbox(filtered: string[]) {
    const { addValues, removeValues, options, value } = this.props;
    if (!addValues || !removeValues) return null;

    const valueSet = typeof value === 'string' || !value ? new Set([value]) : value;
    let checkedCount = 0;
    let indeterminate = false;
    for (let i = 0; i < filtered.length; i++) {
      const match = valueSet.has(filtered[i]);
      if (match) checkedCount++;
      if (checkedCount && checkedCount <= i) {
        indeterminate = true;
        break;
      }
    }
    const checked = Boolean(checkedCount) && checkedCount === filtered.length;
    const title = `Click to ${checked ? 'unselect' : 'select'} all ${
      filtered.length < options.length ? 'filtered ' : ''
    }options`;

    return (
      <Tooltip title={title}>
        <Checkbox
          className="FilteredList--filterCheckbox"
          checked={checked}
          disabled={!filtered.length}
          onChange={({ target: { checked: newCheckedState } }) => {
            if (newCheckedState) addValues(filtered.filter(f => !valueSet.has(f)));
            else removeValues(filtered);
          }}
          indeterminate={indeterminate}
        />
      </Tooltip>
    );
  }

  private getFilteredOptions = () => {
    const { options } = this.props;
    const { filterText } = this.state;
    return filterText ? matchSorter(options, filterText) : options;
  };

  private setValue = (value: string) => {
    this.props.setValue(value);
    this.setState({ filterText: '', focusedIndex: null });
  };

  private onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const { focusedIndex: stFocused } = this.state;
    switch (event.key) {
      case EKey.Escape: {
        const { cancel } = this.props;
        this.setState({ filterText: '', focusedIndex: null });
        if (cancel) cancel();
        break;
      }
      case EKey.ArrowUp:
      case EKey.ArrowDown: {
        const { visibleStartIndex, visibleStopIndex } = this.state;
        let focusedIndex: number | void;
        if (stFocused == null) {
          focusedIndex = event.key === EKey.ArrowDown ? visibleStartIndex : visibleStopIndex;
          this.setState({ focusedIndex });
        } else {
          const offset = event.key === EKey.ArrowDown ? 1 : -1;
          const filteredOptions = this.getFilteredOptions();
          const i = stFocused + offset;
          focusedIndex = i > -1 ? i % filteredOptions.length : filteredOptions.length + i;
          this.setState({ focusedIndex });
        }
        const listInstance = this.vlistRef.current;
        if (listInstance && (focusedIndex < visibleStartIndex + 1 || focusedIndex > visibleStopIndex - 1)) {
          listInstance.scrollToItem(focusedIndex);
        }
        break;
      }
      case EKey.Enter: {
        const filteredOptions = this.getFilteredOptions();
        if (stFocused !== null) this.setValue(filteredOptions[stFocused]);
        else if (filteredOptions.length === 1) this.setValue(filteredOptions[0]);
        break;
      }
      default: // no-op
    }
  };

  private onListScrolled = _debounce((scrollInfo: ListOnScrollProps) => {
    if (!scrollInfo.scrollUpdateWasRequested) {
      this.setState({ focusedIndex: null });
    }
  }, 80);

  private onListItemsRendered = _debounce((viewInfo: ListOnItemsRenderedProps) => {
    const { visibleStartIndex, visibleStopIndex } = viewInfo;
    this.setState({ visibleStartIndex, visibleStopIndex });
  }, 80);

  private onFilterChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ filterText: event.target.value, focusedIndex: null });
  };

  render() {
    const { addValues, multi, options, removeValues, value } = this.props;
    const { filterText, focusedIndex } = this.state;
    const filteredOptions = this.getFilteredOptions();
    const filteredCheckbox = multi && this.getFilteredCheckbox(filteredOptions);
    const data = {
      addValues,
      focusedIndex,
      highlightQuery: filterText,
      multi,
      options: filteredOptions,
      removeValues,
      selectedValue: value,
      setValue: this.setValue,
    };
    return (
      <div ref={this.wrapperRef}>
        <div className="FilteredList--filterWrapper">
          {filteredCheckbox}
          <label className="FilteredList--inputWrapper">
            <IoIosSearch className="FilteredList--filterIcon" />
            <input
              className="FilteredList--filterInput"
              placeholder="Filter..."
              onChange={this.onFilterChanged}
              onKeyDown={this.onKeyDown}
              ref={this.inputRef}
              type="text"
              value={filterText}
            />
          </label>
        </div>
        <VList
          key={filterText}
          className="FilteredList--list u-simple-scrollbars"
          height={Math.min(options.length * ITEM_HEIGHT, MAX_HEIGHT)}
          itemCount={filteredOptions.length}
          itemData={data}
          itemSize={ITEM_HEIGHT}
          width={650}
          overscanCount={25}
          onItemsRendered={this.onListItemsRendered}
          onScroll={this.onListScrolled}
          ref={this.vlistRef}
        >
          {ListItem}
        </VList>
      </div>
    );
  }
}
