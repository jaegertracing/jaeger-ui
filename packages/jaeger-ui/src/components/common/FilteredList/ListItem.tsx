// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { ListChildComponentProps } from 'react-window';

import highlightMatches from './highlightMatches';

import './ListItem.css';

interface IListItemProps extends ListChildComponentProps {
  data: {
    focusedIndex: number | null;
    highlightQuery: string;
    options: string[];
    selectedValue: string | null;
    setValue: (value: string) => void;
  };
}

export default class ListItem extends React.PureComponent<IListItemProps> {
  onClicked = () => {
    const { data, index } = this.props;
    const { options, setValue } = data;
    const value = options[index];
    setValue(value);
  };

  render() {
    const { data, index, style: styleOrig } = this.props;
    // omit the width from the style so the panel can scroll horizontally
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { width: _, ...style } = styleOrig;
    const { focusedIndex, highlightQuery, options, selectedValue } = data;
    const cls = cx('FilteredList--ListItem', {
      'is-focused': index === focusedIndex,
      'is-selected': options[index] === selectedValue,
      'is-striped': index % 2,
    });
    return (
      <div
        className={cls}
        style={style}
        onClick={this.onClicked}
        role="switch"
        aria-checked={index === focusedIndex ? 'true' : 'false'}
      >
        {highlightMatches(highlightQuery, options[index])}
      </div>
    );
  }
}
