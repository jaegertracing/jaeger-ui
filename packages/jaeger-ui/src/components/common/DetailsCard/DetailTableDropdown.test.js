// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Button } from 'antd';

import FilteredList from '../FilteredList';
import DetailTableDropdown from './DetailTableDropdown';

describe('DetailTable', () => {
  const options = ['foo', 'bar', 'baz'];
  const props = {
    clearFilters: jest.fn(),
    confirm: jest.fn(),
    options,
    selectedKeys: options.slice(1),
    setSelectedKeys: jest.fn(),
  };
  let wrapper;

  beforeEach(() => {
    props.clearFilters.mockReset();
    props.confirm.mockReset();
    props.setSelectedKeys.mockReset();
    wrapper = shallow(<DetailTableDropdown {...props} />);
  });

  describe('render', () => {
    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('filters duplicates and numbers out of selectedKeys', () => {
      const dupedKeysWithNumbers = props.selectedKeys
        .concat(props.selectedKeys)
        .concat([4, 8, 15, 16, 23, 42]);
      wrapper.setProps({ selectedKeys: dupedKeysWithNumbers });
      expect(wrapper.find(FilteredList).prop('value')).toEqual(new Set(props.selectedKeys));
    });

    it('handles missing clearFilters prop', () => {
      wrapper.setProps({ clearFilters: undefined });
      expect(() =>
        wrapper
          .find(Button)
          .first()
          .simulate('click')
      ).not.toThrow();
    });
  });

  describe('cancel', () => {
    const selectedKeys = [options[0]];

    it('resets to this.selectedKeys on cancel and calls confirm once props reflect cancellation', () => {
      wrapper.instance().selected = selectedKeys;
      expect(props.confirm).not.toHaveBeenCalled();
      expect(props.setSelectedKeys).not.toHaveBeenCalled();

      wrapper
        .find(Button)
        .at(1)
        .simulate('click');
      expect(props.setSelectedKeys).toHaveBeenCalledTimes(1);
      expect(props.setSelectedKeys).toHaveBeenCalledWith(selectedKeys);
      expect(props.confirm).not.toHaveBeenCalled();

      wrapper.setProps({ selectedKeys });
      expect(props.setSelectedKeys).toHaveBeenCalledTimes(1);
      expect(props.confirm).toHaveBeenCalledTimes(1);
    });

    it('updates this.selectedKeys on open/close', () => {
      expect(wrapper.instance().selected).not.toEqual(selectedKeys);

      wrapper.setProps({ selectedKeys: selectedKeys.slice() });
      expect(wrapper.instance().selected).not.toEqual(selectedKeys);

      wrapper.setProps({ selectedKeys: selectedKeys.slice() });
      expect(wrapper.instance().selected).toEqual(selectedKeys);
    });

    it('maintains this.selectedKeys on changed selection', () => {
      wrapper.instance().selected = selectedKeys;
      wrapper.setProps({ selectedKeys: props.options.slice(0, props.selectedKeys.length) });
      expect(wrapper.instance().selected).toBe(selectedKeys);
    });
  });

  describe('FilteredList interactions', () => {
    const getFn = propName => wrapper.find(FilteredList).prop(propName);

    it('adds values', () => {
      const newValues = props.options.map(o => `not-${o}`);
      getFn('addValues')(newValues);
      expect(props.setSelectedKeys).toHaveBeenCalledWith([...props.selectedKeys, ...newValues]);
    });

    it('removes values', () => {
      getFn('removeValues')([props.selectedKeys[0]]);
      expect(props.setSelectedKeys).toHaveBeenCalledWith(props.selectedKeys.slice(1));
    });

    it('sets a value', () => {
      getFn('setValue')(props.options[0]);
      expect(props.setSelectedKeys).toHaveBeenCalledWith([props.options[0]]);
    });
  });
});
