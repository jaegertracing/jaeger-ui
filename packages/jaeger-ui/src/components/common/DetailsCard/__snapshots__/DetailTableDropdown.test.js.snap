// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailTable render renders as expected 1`] = `
<div>
  <FilteredList
    addValues={[Function]}
    multi={true}
    options={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    removeValues={[Function]}
    setValue={[Function]}
    value={
      Set {
        "bar",
        "baz",
      }
    }
  />
  <div
    className="DetailTableDropdown--Footer"
  >
    <Tooltip
      overlayClassName="DetailTableDropdown--Tooltip"
      title="Remove filter from this column"
    >
      <Button
        className="DetailTableDropdown--Btn Clear"
        onClick={[MockFunction]}
      >
        <IoTrash
          size={18}
        />
        Clear Filter
      </Button>
    </Tooltip>
    <div
      className="DetailTableDropdown--Footer--CancelConfirm"
    >
      <Tooltip
        overlayClassName="DetailTableDropdown--Tooltip"
        title="Cancel changes to this column's filter"
      >
        <Button
          className="DetailTableDropdown--Btn Cancel"
          onClick={[Function]}
        >
          <IoBan
            size={20}
          />
          Cancel
        </Button>
      </Tooltip>
      <Tooltip
        overlayClassName="DetailTableDropdown--Tooltip"
        title={
          <div
            className="DetailTableDropdown--Tooltip--Body"
          >
            <span>
              Apply changes to this column's filter
            </span>
            <span>
              Same effect as clicking outside the dropdown
            </span>
          </div>
        }
      >
        <Button
          className="DetailTableDropdown--Btn Apply"
          onClick={[MockFunction]}
        >
          <IoCheckmark
            size={18}
          />
          Apply
        </Button>
      </Tooltip>
    </div>
  </div>
</div>
`;
